Combiners

AndyMcAliley opened the pull request:
- Using `tar_combine()`, ombine tallies of observations per year, and use them to plot data coverage through time.
- Summarize time series plots of discharge at oldest sites to track changes.
- Create an interactive map showing site locations and time series plots.

Here's a screenshot of the interactive plot:

<img width="1148" alt="Screen Shot 2021-09-27 at 4 04 15 PM" src="https://user-images.githubusercontent.com/11847289/134992018-1e142cbb-e1fa-4013-bce1-b3fb3386f54b.png">

There's one odd behavior: if I add/remove a state, sometimes every target gets rebuilt. If I add/remove a state again right away, though, then only the expected targets are remade. It doesn't seem to matter whether I'm adding or removing a state. It seems like this should only happen when the `nwis_inventory_XX` targets change, so I'm not sure why this happens. 

See #10

wmcaliley-usgs committed:
Combine tallies and plot data coverage

wmcaliley-usgs committed:
Track summary of state timeseries plots

wmcaliley-usgs committed:
Add interactive leaflet map of state timeseries

hcorson-dosch-usgs commented on _targets.R:
```diff
@@ -6,43 +6,67 @@ suppressPackageStartupMessages(library(dplyr))
+  # Pull site data, tally, and plot
+  mapped_by_state_targets,
+  # Combine tallies
+  tar_combine(obs_tallies, mapped_by_state_targets[[3]], command = combine_obs_tallies(!!!.x)),
```
Just a note that you could also use `mapped_by_state_targets$tally` here. It makes no difference functionally, but I prefer it because it allows me to see visually which of the mapped targets are being combined in this step, rather than having to remember which # is assigned to each of the mapped targets.

hcorson-dosch-usgs commented on _targets.R:
```diff
@@ -6,43 +6,67 @@ suppressPackageStartupMessages(library(dplyr))
+  tar_combine(obs_tallies, mapped_by_state_targets[[3]], command = combine_obs_tallies(!!!.x)),
+  # Summarize targets
+  tar_combine(summary_state_timeseries_csv, 
+              mapped_by_state_targets[[4]], 
```
And this could be `mapped_by_state_targets$timeseries_png`

hcorson-dosch-usgs commented on _targets.R:
```diff
@@ -6,43 +6,67 @@ suppressPackageStartupMessages(library(dplyr))
+             map_sites("3_visualize/out/site_map.png", oldest_active_sites), 
+             format = "file"),
+  # Interactive map of oldest sites, with timeseries plots
+  tar_target(timeseries_map_html,
```
Nice work remembering to add this `_html` suffix to the target name!

hcorson-dosch-usgs commented:
> There's one odd behavior: if I add/remove a state, sometimes every target gets rebuilt. If I add/remove a state again right away, though, then only the expected targets are remade. It doesn't seem to matter whether I'm adding or removing a state. It seems like this should only happen when the `nwis_inventory_XX` targets change, so I'm not sure why this happens.

Hmm, this does sound odd. Were you switching between [these two vectors](https://github.com/AndyMcAliley/ds-pipelines-targets-3/blob/combiners/_targets.R#L23-L24) to test this? I'm not at all certain, but I wonder if `targets` saw that uncommenting/commenting out as a change to the code and an altogether new `states` vector, since it's on a different line. If that was what you were doing, what happens if you just add or remove a state in line 24, without uncommenting out line 23? If not - I can ping Lindsay tomorrow and run this by her.

hcorson-dosch-usgs reviewed (approved):
Everything looks great, Andy! Let's just get to the bottom of that odd rebuilding issue before you merge. See my comment, and if that doesn't address it, let's run it by Lindsay tomorrow.

AndyMcAliley commented:
> Hmm, this does sound odd. Were you switching between [these two vectors](https://github.com/AndyMcAliley/ds-pipelines-targets-3/blob/combiners/_targets.R#L23-L24) to test this? 

Yes, I was!
> What happens if you just add or remove a state in line 24, without uncommenting out line 23? 

I'm still seeing the same behavior:
- I ran `tar_make()` once this morning. Every target was skipped because I hadn't made any changes.
- I removed `IA` from line 24 and ran `tar_make()`. Every target was remade.
- I added `IA` back into line 24. Running `tar_make()` only remade the expected targets, skipping `nwis_data_XX`, `time series_png_XX`, and `tally_XX` targets for all states other than `IA`. 
- I removed `IA` from line 24 again. Running `tar_make()` only remade the expected targets.
- I added `IA` back into line 24. Running `tar_make()` only remade the expected targets. 
- I restarted RStudio.
- I ran `library(targets)` and `tar_make()`. Every target was skipped.
- I removed `IA` from line 24. Running `tar_make()` only remade the expected targets.

AndyMcAliley commented:
- I thought, "Surely this is enough testing" and commented.
- I added `IA` back into line 24 to end with all 6 states. 
- Running tar_make() skipped `nwis_data_MI`, `nwis_data_MN`, `nwis_data_IN`, `nwis_data_IA`, and `nwis_data_WI`. `nwis_data_IL` was remade. What? `IA` should have been remade, not `IL`!

I wonder if the download performed by `find_oldest_site` subtly changes at times? Maybe the next troubleshooting step is to figure out why certain targets are remade. Is there a quick way to compare old and new versions of targets or figure out why `targets` treats a target as having changed?

hcorson-dosch-usgs commented:
Ah, ok. I think you've hit on part of the explanation here w/ the changes to `oldest_active_sites`. I went back to the `find_oldest_site()` function, and it looks like it is setting the `startDt` and `endDt` in the nwis call based on the system time, so that will change every single calendar day. So when you first ran `tar_make()` this morning, before making any changes, everything was indeed up to date, but then your removal of `IA` from the `states` vector caused `targets` to rebuild `oldest_active_sites`, and because the system date today is one day later than what it was yesterday, the `end_date` column values changed for every row of `oldest_active_sites` and therefore each state's `nwis_inventory_XX` object, which triggered rebuilds of the `nwis_data_XX` targets. That's why it triggered a full-pipeline rebuild of every target. In your next step, however, adding `IA` had no impact on the `nwis_inventory_XX` for other states, which meant only the IA-specific `nwis_data_IA`, `timeseries_png_IA` and `tally_IA` were affected.

As for your last bullet - I think you're hitting on some very unique `targets` case scenarios:
> I added IA back into line 24 to end with all 6 states. Running tar_make() skipped nwis_data_MI, nwis_data_MN, nwis_data_IN, nwis_data_IA, and nwis_data_WI. nwis_data_IL was remade. What? IA should have been remade, not IL!

Maybe something did change for the IL row when `oldest_active_sites` was built? The `count_nu` column, maybe?

And for IA - between pipeline builds, `targets` memory isn't wiped -- which means that even though you removed `IA` from the `states` vector and had `targets` rebuild the pipeline and remove `IA` from the downstream summary targets  like `obs_tallies`, `data_coverage_png` and `timeseries_map_html`, `targets` is still retaining `nwis_inventory_IA`, `nwis_data_IA`, `tally_IA`, and `timeseries_IA` as objects in the `_targets/objects` directory, and `timeseries_IA.png` still exists in `3_visualize/out`. 

This is my _theory_, based on testing this in my own pipelines-III repo, but @lindsayplatt  I would welcome your thoughts here: When we (in quick succession, on the same calendar day) 1) build the full pipeline with `IA`,  2) remove `IA` from the `states` vector, 3) rebuild, 4) add it back in, and 5) then rebuild _again_, `targets` rebuilds `oldest_active_sites` to include `IA`, and (here is where it gets weird and I'm just hypothesizing) sees that `oldest_active_sites` is unchanged _relative to when_ `nwis_inventory_IA` (which still exists in the `_targets/objects` directory) was built, so **skips that target** (but it _does_ rebuild the `nwis_inventory_XX` targets for all other states, b/c `oldest_active_sites` has changed since they were built before in step 3). Then, because `nwis_inventory_IA` is unchanged, and `targets` still has these IA-specific targets in `_targets/objects`, it doesn't rebuild `nwis_data_IA`, `tally_IA` or `timeseries_png_IA`:
![image](https://user-images.githubusercontent.com/54007288/135172277-66d44e1d-7d18-461d-8169-79458644939c.png)


Typically we're not adding and removing the same parameter over and over, but if you were and for some reason needed to wipe `targets`'s memory, you could run `tar_destroy()` (note: use tar_destroy() very sparingly and deliberately, as it wipes **all previous pipeline builds**). That will wipe all of those stored objects, _but not_ any already saved `*/out/*` files.
 
 
 ______________________________________________________________________________________________________

So I think that leaves us with these responses to the behavior you saw:

> I ran tar_make() once this morning. Every target was skipped because I hadn't made any changes.

Correct. No changes were made, and the pipeline was already built --> all targets were skipped

> I removed IA from line 24 and ran tar_make(). Every target was remade.

Yes, b/c the removal of `IA` changed the `states` vector, which triggered a rebuild of `oldest_active_sites`, and since the system day was now different, `nwis_inventory_XX` changed for each state, triggering a rebuild of the targets for all states.

> I added IA back into line 24. Running tar_make() only remade the expected targets, skipping nwis_data_XX, time series_png_XX, and tally_XX targets for all states other than IA.

Right. At this point, the `IA`-specific targets had not yet been built since the system date had changed, triggering that aforementioned change to each of the `nwis_inventory_XX` targets, so the `nwis_data_IA` target and downstream targets all had to be built.

> I removed IA from line 24 again. Running tar_make() only remade the expected targets.

Right b/c nothing had changed for the states remaining in the `states` vector

> I added IA back into line 24. Running tar_make() only remade the expected targets.

Here's where I'd _expect_ you actually wouldn't see a rebuild of the IA-specific targets, given what I noted above re: `targets` memory. Possibly, if you did see the rebuild of those IA targets, something _had_ changed in the IA row of `oldest_active_sites`, as might have been the case for IL in your final bullet?

> I restarted RStudio. I ran library(targets) and tar_make(). Every target was skipped.

Right, b/c everything had been built already and you had made no changes

> I removed IA from line 24. Running tar_make() only remade the expected targets.

Right b/c nothing had changed for the states remaining in the `states` vector

> I added IA back into line 24 to end with all 6 states. Running tar_make() skipped nwis_data_MI, nwis_data_MN, nwis_data_IN, nwis_data_IA, and nwis_data_WI. nwis_data_IL was remade. What? IA should have been remade, not IL!

And see my notes above about what I think was happening in this step.

AndyMcAliley commented:
Wow! Thank you, Hayley! That really helps - and sorry to send you into all that sleuthing! It makes complete sense that the daily changes to `startDt` and `endDt` would trigger rebuilds each morning. 

And you're right about `IA` not being rebuilt when it's added _back_ in. I wonder if I glazed over that a couple of days ago... or maybe, like you suggested, something in the `IA` row of `oldest_active_sites` happened to change, similar to what may have happened with `IL`. I appreciate your explanation of why `IA` is not rebuilt when everything else is - that also makes sense that `oldest_active_sites` ends up being identical to what it was when the `IA` objects were last built.

I really appreciate all the feedback!
